File: Sorting/Merge.c
#include <stdio.h>

void merge(int *a, int l, int mid, int h)
{
    int i = l;
    int j = mid + 1;
    int temp[10];
    int k = 0;
    int s = 0;

    while (i <= mid && j <= h)
    {
        if (a[i] < a[j])
        {
            temp[k] = a[i];
            k++;
            i++;
        }
        else
        {
            temp[k] = a[j];
            k++;
            j++;
        }
    }
    if (i > mid)
    {
        for (; j <= h; j++)
        {
            temp[k] = a[j];
            k++;
        }
    }
    if (j > h)
    {
        for (; i <= mid; i++)
        {
            temp[k] = a[i];
            k++;
        }
    }

    for (s = l; s <= h; s++)
    {
        a[s] = temp[s];
    }
}
void mergeSort(int *a, int l, int h)
{
    int mid = 0;

    if (l < h)
    {
        mid = (l + h) / 2;
        mergeSort(a, l, mid);
        mergeSort(a, mid + 1, h);
        merge(a, l, mid, h);
    }
    return;
}
int main()
{
    int arr[10] = {23, 34, 29, 2, 24, 9, 3, 8, 39, 7};
    mergeSort(arr, 0, 9);
    for (int i = 0; i < 10; i++)
    {
        printf("%d \n", arr[i]);
    }
}

File: Sorting/Merge.java
public class Merge {

    public static void mergea(int[] arr, int mid, int l, int h) {
        int n1 = (mid - l) + 1;
        int n2 = h - mid;
        int[] L = new int[n1];
        int[] R = new int[n2];

        for (int i = 0; i < n1; i++) {
            L[i] = arr[l + i];
        }
        for (int j = 0; j < n2; j++) {
            R[j] = arr[mid + 1 + j];
        }

        int i = 0;
        int j = 0;

        int k = l;
        while (i < n1 && j < n2) {
            if (L[i] < R[j]) {
                arr[k] = L[i];
                k++;
                i++;
            } else {

                arr[k] = R[j];
                k++;
                j++;

            }
        }

        while (i < n1) {
            arr[k] = L[i];
            k++;
            i++;
        }

        while (j < n2) {
            arr[k] = R[j];
            k++;
            i++;
        }

    }

    public static void mergeSort(int[] arr, int low, int high) {
        int mid = 0;
        if (low < high) {

            mid = low + (high - low) / 2;

            mergeSort(arr, low, mid);
            mergeSort(arr, mid + 1, high);
            mergea(arr, mid, low, high);
        }
        return;
    }

    public static void main(String[] args) {
        int[] arr = { 23, 4, 1, 6, 3, 4 };
        mergeSort(arr, 0, (arr.length - 1));
        for (int x : arr) {
            System.out.print(x + " ");
        }

    }
}
